### PR TITLE
move BlockNumberProvider

### PR DESCRIPTION
**Pull request by xlc**
move BlockNumberProvider from `sp_runtime::offchain::storage_lock` to `sp_runtime::traits`

Useful for https://github.com/paritytech/cumulus/issues/514

### CODE REVIEW

**Review by coriolinus — APPROVED** [review]

File: primitives/runtime/src/traits.rs [hunk 1]
```diff
@@ -1460,6 +1460,29 @@ pub trait BlockIdTo<Block: self::Block> {
...
+pub trait BlockNumberProvider {
+	/// Type of `BlockNumber` to provide.
+	type BlockNumber: Codec + Clone + Ord + Eq + AtLeast32BitUnsigned;
+
+	/// Returns the current block number.
+	///
+	/// Provides an abstraction over an arbitrary way of providing the
+	/// current block number.
+	///
+	/// In case of using crate `sp_runtime` without the crate `frame`
```
Comment (bkchr) [inline primitives/runtime/src/traits.rs:1473]:
```suggestion
	/// In case of using crate `sp_runtime` with the crate `frame`
```

**Review by bkchr — APPROVED** [review]

File: primitives/runtime/src/traits.rs [hunk 2]
```diff
@@ -1460,6 +1460,29 @@ pub trait BlockIdTo<Block: self::Block> {
...
+	/// system, it is already implemented for
```
Comment (bkchr) [inline primitives/runtime/src/traits.rs:1474]:
```suggestion
	/// it is already implemented for
```

**Comment by bkchr** [comment]
bot merge

**Comment by JoshOrndorff** [comment]
Does this need a Polkadot companion?

**Comment by kianenigma** [comment]
bot merge

**Pull request merged**